### PR TITLE
feat(abstract-eth): refactor eth wrw recovery tss

### PR DESCRIPTION
**Pull request by dpkjnr**
Ticket: WP-1885

### CODE REVIEW

**Review by mullapudipruthvik — COMMENTED** [review]
do we have a test for this?, so what we know the refactor didn't break anything

**Comment by dpkjnr** [comment]
> do we have a test for this?, so what we know the refactor didn't break anything

gg18 recovery tests are already there and those are passing. I will add the DKLS specific tests with DKLS recovery signing

**Review by mullapudipruthvik — APPROVED** [review]

File: modules/abstract-eth/src/abstractEthLikeNewCoins.ts [hunk 1]
```diff
@@ -1902,56 +1902,79 @@ export abstract class AbstractEthLikeNewCoins extends AbstractEthLikeCoin {
...
+        signingMaterial.bitgoNShare &&
+        (signingMaterial.userNShare || signingMaterial.backupNShare)
+      );
+    } catch (error) {
+      return false;
+    }
+  }
 
-    // build recipients object
+  private async buildTssRecoveryTxn(baseAddress: string, gasPrice: any, gasLimit: any, params: RecoverOptions) {
```
Comment (sachushaji) [inline modules/abstract-eth/src/abstractEthLikeNewCoins.ts:1975]: NIT: we could type gasPrice and gaslimit

File: modules/abstract-eth/src/abstractEthLikeNewCoins.ts [hunk 2]
```diff
@@ -1966,44 +1989,35 @@ export abstract class AbstractEthLikeNewCoins extends AbstractEthLikeCoin {
...
-        params.replayProtectionOptions
-      );
-    }
+  async validateBalanceAndGetTxAmount(baseAddress: string, gasPrice: BN, gasLimit: BN) {
+    const baseAddressBalance = await this.queryAddressBalance(baseAddress);
 
-    const signableHex = tx.getMessageToSign(false).toString('hex');
-    if (!userKeyCombined || !backupKeyCombined) {
-      throw new Error('Missing key combined shares for user or backup');
+    const totalGasNeeded = gasPrice.mul(gasLimit);
```
Comment (sachushaji) [inline modules/abstract-eth/src/abstractEthLikeNewCoins.ts:2009]:
```suggestion
    const totalGasCostNeeded = gasPrice.mul(gasLimit);
```
NIT: Might be the correct term

File: modules/abstract-eth/src/abstractEthLikeNewCoins.ts [hunk 3]
```diff
@@ -1966,44 +1989,35 @@ export abstract class AbstractEthLikeNewCoins extends AbstractEthLikeCoin {
...
+    const weiToGwei = new BN(10 ** 9);
+    if (baseAddressBalance.lt(totalGasNeeded)) {
+      throw new Error(
+        `Backup key address ${baseAddress} has balance ${baseAddressBalance.div(weiToGwei).toString()} Gwei.` +
```
Comment (sachushaji) [inline modules/abstract-eth/src/abstractEthLikeNewCoins.ts:2013]: There is no concept of a backup key address in GG18, all the public keys are corresponding to the base address. So we could change this error message to say base address and not backup address which could confuse people

File: modules/abstract-eth/src/abstractEthLikeNewCoins.ts (hunk 1 above)
Comment (dpkjnr) [reply modules/abstract-eth/src/abstractEthLikeNewCoins.ts:1975]: will make this change along with the DKLS signing PR

File: modules/abstract-eth/src/abstractEthLikeNewCoins.ts (hunk 2 above)
Comment (dpkjnr) [reply modules/abstract-eth/src/abstractEthLikeNewCoins.ts:2009]: will make this change along with the DKLS signing PR

File: modules/abstract-eth/src/abstractEthLikeNewCoins.ts (hunk 3 above)
Comment (dpkjnr) [reply modules/abstract-eth/src/abstractEthLikeNewCoins.ts:2013]: will make this change along with the DKLS signing PR

**Pull request merged**